[lld][ELF] Add --why-live flag (inspired by Mach-O)

mysterymath opened the pull request:
This prints a stack of reasons that symbols that match the given glob(s) survived GC. It has no effect unless section GC occurs.

This implementation does not require -ffunction-sections or -fdata-sections to produce readable results, althought it does tend to work better (as does GC).

Details about the semantics:
 - Some chain of liveness reasons is reported; it isn't specified which chain.
 - A symbol or section may be live:
   - Intrisically (e.g., entry point)
   - Because needed by a live symbol or section
   - (Symbols only) Because part of a section live for another reason
   - (Sections only) Because they contain a live symbol
 - Both global and local symbols (`STB_LOCAL`) are supported.
 - References to symbol + offset are considered to point to:
   - If the referenced symbol is a section (`STT_SECTION`):
     - If a sized symbol encloses the referenced offset, the enclosing symbol.
     - Otherwise, the section itself, generically.
   - Otherwise, the referenced symbol.

mysterymath committed:
[lld][ELF] Add --why-live flag (inspired by Mach-O)

This prints the stack of reasons that symbols that match the given
glob(s) survived GC. It has no effect unless section GC occurs.

A symbol may be live intrisically, because referenced by another symbol
or section, or because part of a live section. Sections have similar
reasons.

This implementation does not require -ffunction-sections or
-fdata-sections to produce readable results, althought it does tend to
work better (as does GC).

MaskRay commented on lld/ELF/MarkLive.cpp:
```diff
@@ -101,6 +116,12 @@ void MarkLive<ELFT>::resolveReloc(InputSectionBase &sec, RelTy &rel,
 
+  LiveReason reason;
+  if (!ctx.arg.whyLive.empty()) {
+    Defined *reasonSym = sec.getEnclosingSymbol(rel.r_offset);
```
getEnclosingSymbol iterates all symbols in the file, which is very slow. This perhaps should only be run after we get a patch from a root to the --why-live specified symbol.

MaskRay reviewed (commented):
--gc-sections might consume 10% of the total link time. The performance is important. Would the two extra arguments slow down the pass too much?

MaskRay commented on lld/test/ELF/why-live.s:
```diff
@@ -0,0 +1,132 @@
+.globl _start
+.section ._start,"ax",@progbits
+_start:
+jmp test_simple
```
indent instructions but keep labels unintended

MaskRay commented on lld/ELF/MarkLive.cpp:
```diff
@@ -201,15 +235,71 @@ void MarkLive<ELFT>::enqueue(InputSectionBase *sec, uint64_t offset) {
+  }
+
+  auto msg = Msg(ctx);
+  msg << "live symbol: " << toStr(ctx, *s);
```
`<< s`

(`toStr` should be avoided if feasible)

smithp35 commented:
Apologies for the delay in responding. I've just started looking through the code. 

I think it would be helpful to update the description with a mini specification of what you intend the output to be. There will have been some design decisions made on what to include in the output. At the moment this needs to be reverse engineered out of the code and tests.

For example:
* Do we report all possible chains of reason for each symbol for each reloc or just the first one we find? 
* If we include all possible chains do we attempt to filter out chains that have the same parent section (.text.1 has multiple relocations to .text.2)?

smithp35 reviewed (commented):
Thanks for working on this. I've got some subjective suggestions about the external interface.
* I recommend adding the object and section information as well as the symbol (possibly via a verbose mode). This particularly useful when there are multiple symbols and sections with the same name.
* For the root symbol, I recommend adding information about the source if its easily available. For example _start is the entry point. This could be particularly useful if a symbol is unexpectedly made live via a linker script reference.

Finally, one that you may want to exclude, is any Arm/AArch64 mapping symbol from being included in the why-live output. These often are defined at the same address as function and data symbols, but you'll almost certainly want to prefer the function and data symbol.

mysterymath commented:
> For example:
> 
> * Do we report all possible chains of reason for each symbol for each reloc or just the first one we find? We could prefer the shortest path.

An unspecified path (the first one we find); updated the description accordingly. When the user reaches for `--why-live`, since they're asking "why is this thing surviving GC?", it's very likely that something shouldn't have survived GC. In this case, any chain to that thing is erroneous, so any chain we report will help the user. If they break chain link, it still survive, but in that case they just need to re-run the link with `--why-live` (and they need to do this anyway to verify the fix). Reporting all chains may save some steps, but it could complicate the output; we would need extra effort to keep cardinalities small and to keep the output readable.

> * If we include all possible chains do we attempt to filter out chains that have the same parent section (.text.1 has multiple relocations to .text.2)?

See above.

> * What do we report for each chain? The output looks like it tries to find an enclosing function, but that could be ambiguous if it were a STB_LOCAL function, perhaps always worth adding the file and section?

See below.

> * Does this apply to just prevailing global symbol definitions, or can it be local symbols too? If local symbols are permitted does this apply to all local symbols that match the name?

I'm just scanning the linker's symbol table, so IIUC that shouldn't include `STB_LOCAL` symbols. It may be valuable to add support for this later, but it does create the ambiguity concerns that you bring up above. 

> * For relocations to section-symbol+offset can a user refer to a STT_FUNC or STT_OBJECT symbol defined at that location and expect to use that as an input to why-live?

Yes, that works. If there is an enclosing sized symbol at that location, then that is reported. Otherwise, the reference is considered a generic reference to the section. This case is tested by `test_section_offset`. Updated the description.

> * For chains of reasons that include section-symbol+offset relocations, is the linker expected to print a STT_FUNC or STT_OBJECT symbol defined at the same offset in the chain of reasons or is section-symbol+offset preferred?

Yes, enclosing objects are preferred. Updated the description.

> * Can I ask for why a section rather than a symbol is live? For example I may have a section that defines many symbols and I want some way of saying why-live=all-symbols-defined-in-section. This would need something like an InputSectionDescription to identify which section.

Not at the moment. One can do this by naming any one global symbol defined in the section. If such a symbol is directly referenced, then that reference will be reported. Otherwise, it will be reported alive by virtue of being part of the section, and some reference to the section will be reported as the reason the section is alive.

> * Do we report the source of any explicit entry points such as KEEP in a linker script, or a command line option? The user should be aware of what they've kept, but may not be aware of the implications. Perhaps there's an opportunity for a --what-live section, this would say what sections are being made live as a result of it being live.

Not at present.

mysterymath commented:
Actually, I think I was wrong about `STB_LOCAL`. The `whyLive` data structure records and may report any symbol used in a reference, and some of these may be `STB_LOCAL`. It's just the top-level scan that looks through the global symbol table. So we'd need a way to disambiguate local symbols regardless, and that means we may as well scan the files themselves. I'll take this as a TODO for this patch.

mysterymath commented:
> * For the root symbol, I recommend adding information about the source if its easily available. For example _start is the entry point. This could be particularly useful if a symbol is unexpectedly made live via a linker script reference.

I think this should be fairly straightforward to add, if annoying to test.
 
> Finally, one that you may want to exclude, is any Arm/AArch64 mapping symbol from being included in the why-live output. These often are defined at the same address as function and data symbols, but you'll almost certainly want to prefer the function and data symbol.

Only sized or directly referenced symbols can be reported as liveness reasons. Given that the mapping symbols mark positions, it I'd assume they wouldn't have either of those characteristics. Is that correct?

MaskRay reviewed (commented):
I think we need Mach-O style `template <bool RecordWhyLive>` to prevent overhead with the regular --gc-sections code path.

https://reviews.llvm.org/D69607 is a previous attempt adding the option.

smithp35 commented:
> > Finally, one that you may want to exclude, is any Arm/AArch64 mapping symbol from being included in the why-live output. These often are defined at the same address as function and data symbols, but you'll almost certainly want to prefer the function and data symbol.
> 
> Only sized or directly referenced symbols can be reported as liveness reasons. Given that the mapping symbols mark positions, I'd assume they wouldn't have either of those characteristics. Is that correct?

Yes that is correct. They aren't directly called and they have no size.

mysterymath commented:
> I think we need Mach-O style `template <bool RecordWhyLive>` to prevent overhead with the regular --gc-sections code path.
> 
> https://reviews.llvm.org/D69607 is a previous attempt adding the option.

It seems prudent to collect some evidence that this might be a performance problem before moving to a less readable implementation. I'm planning to do a simple link of clang with -ffunction-sections and -fdata-sections; that should provide some cursory evidence. If that's not sufficient, I'm also totally okay with this patch being summarily reverted in case a regression occurs in something larger. Especially since it should be entirely mechanical to move this to use `template <bool RecordWhyLive>`.

mysterymath commented:
> > I think we need Mach-O style `template <bool RecordWhyLive>` to prevent overhead with the regular --gc-sections code path.
> > https://reviews.llvm.org/D69607 is a previous attempt adding the option.
> 
> It seems prudent to collect some evidence that this might be a performance problem before moving to a less readable implementation. I'm planning to do a simple link of clang with `-ffunction-sections` and `-fdata-sections`; that should provide some cursory evidence. If that's not sufficient, I'm also totally okay with this patch being summarily reverted in case a regression occurs in something larger. Especially since it should be entirely mechanical to move this to use `template <bool RecordWhyLive>`.

I finally got back around to this, and I tried the above link of clang. Running a Student's t test on 5 trials yields a mean regression of 1.5%, and a very low p value for the means being the same (0.00024). That's definitely way too much, so I'll go through and add the templates. I'm surprised it was this big of a difference, but I'm still glad I measured.

mysterymath committed:
Resolve performance issue by templating MarkLive on --why-live

mysterymath commented:
> > > I think we need Mach-O style `template <bool RecordWhyLive>` to prevent overhead with the regular --gc-sections code path.
> > > https://reviews.llvm.org/D69607 is a previous attempt adding the option.
> > 
> > 
> > It seems prudent to collect some evidence that this might be a performance problem before moving to a less readable implementation. I'm planning to do a simple link of clang with `-ffunction-sections` and `-fdata-sections`; that should provide some cursory evidence. If that's not sufficient, I'm also totally okay with this patch being summarily reverted in case a regression occurs in something larger. Especially since it should be entirely mechanical to move this to use `template <bool RecordWhyLive>`.
> 
> I finally got back around to this, and I tried the above link of clang. Running a Student's t test on 5 trials yields a mean regression of 1.5%, and a very low p value for the means being the same (0.00024). That's definitely way too much, so I'll go through and add the templates. I'm surprised it was this big of a difference, but I'm still glad I measured.

I've upload a performance fix using a TrackWhyLive template. I've verified that the compiler is able to reason that the extra args aren't used in the version of MarkLive where why-live isn't tracked, and 10 trials of the clang link shows no statistically significant difference in performance.

mysterymath committed:
Handle local symbols etc.

- Explicitly track when explicitly live
  - This prevents e.g. _start from being reported live as part of .start.
- Use quad for more control over reloc offsets

mysterymath commented:
I've updated this patch to handle local symbols, but noticed something surprising as I was doing so.

mysterymath committed:
Address various review comments

- Prefer "using"
- Defer lookup of sec+offset liveness reasons to printing time
- Indent instructions
- Avoid unnecessary toStr()
- Provide description strings to print with liveness reasons

mysterymath reviewed (commented):
I've responded to all the review feedback and updated the PR accordingly; PTAL.

mysterymath committed:
clang-format fix

smithp35 commented:
Will take a look this week. Apologies for the delay.

smithp35 commented on lld/ELF/MarkLive.cpp:
```diff
@@ -119,17 +142,29 @@ void MarkLive<ELFT>::resolveReloc(InputSectionBase &sec, RelTy &rel,
 
   for (InputSectionBase *sec : cNamedSections.lookup(sym.getName()))
-    enqueue(sec, 0);
+    enqueue(sec, 0, nullptr, reason);
```
I found the nullptr and std::nullopt callsites a bit more difficult to reason about, I had to keep referring back to the definition to remember what it meant.

Just thinking if there's an easy way to help out. One way would be to make a constant for nullptr like noSym and noObj (or noItem). Possibly a comment like nullptr /* no symbol */. This isn't a strong opinion.

smithp35 commented on lld/ELF/Options.td:
```diff
@@ -559,6 +559,12 @@ defm wrap : Eq<"wrap", "Redirect symbol references to __wrap_symbol and "
                        "__real_symbol references to symbol">,
             MetaVarName<"<symbol>">;
 
+defm why_live
```
I think this will need adding to docs/ld.lld.1 too.

smithp35 commented on lld/ELF/MarkLive.cpp:
```diff
@@ -201,45 +238,126 @@ void MarkLive<ELFT>::enqueue(InputSectionBase *sec, uint64_t offset) {
+  auto msg = Msg(ctx);
+
+  const auto printSymbol = [&](Symbol *s) {
+    if (s->isLocal())
```
It may be worth printing the file for globals as well. It isn't needed to disambiguate the function, but it means I have to search in my source code for it if I want to find it.

Not a strong opinion at first as I expect we can introduce some form of user policy for how much information to print if it is requested by users.

smithp35 reviewed (commented):
Thanks for the update. Some small suggestions from me, but otherwise looks good.

mysterymath committed:
Address review feedback

- LiveObject -> LiveItem
- nit: alive -> live
- Prefer "keeps live" to "(may) make live"
- Inline getParentSec lambda
- Always print symbol file
- reason.item guard clause
- Use /*offset=*/ and /*sym=*/ for enqueue
- Update ld.lld.1

mysterymath commented on lld/ELF/MarkLive.cpp:
```diff
@@ -42,16 +44,30 @@ using namespace lld;
+  std::optional<LiveObject> obj;
+  StringRef desc;
+};
+
```
Acknowledged. I played around with this a bit, but I wasn't able to come up with something without a readability hit. I think there's things we can do if we end up with code size issues in lld overall, but otherwise, the presence of the `TrackWhyLive==true` variants shouldn't upset instruction cache, since they'll be essentially dead for most links.

mysterymath committed:
A few nits

MaskRay commented on lld/ELF/MarkLive.cpp:
```diff
@@ -93,14 +112,18 @@ static uint64_t getAddend(Ctx &, InputSectionBase &sec,
   sym.used = true;
 
+  LiveReason reason;
+  if (TrackWhyLive)
```
`if constexpr `

MaskRay commented on lld/test/ELF/why-live.s:
```diff
@@ -0,0 +1,153 @@
+# FROM-UNSIZED-NEXT: >>> referenced by: {{.*}}.o:(.test_simple)
+# FROM-UNSIZED-NEXT: >>> contained live symbol: {{.*}}.o:(test_simple)
+# FROM-UNSIZED-NEXT: >>> referenced by: {{.*}}.o:(_start) (entry point)
+# FROM-UNSIZED-NOT:  >>>
```
could replace these `-NOT: >>>` with `-EMPTY:`

mysterymath replied:
Done.

smithp35 reviewed (approved):
Thanks for the update. I don't have any more comments. I've marked as approved, but please wait for MaskRay.

mysterymath committed:
Address review comments

- Add comments clarifying invariants and getEnclosinSymbol rationale
- Remove extraneous "-" from ld.lld.1
- Replace "-NOT: >>>" with "-EMPTY:"
- Use split-file to simplify paths

MaskRay commented on lld/ELF/MarkLive.cpp:
```diff
@@ -42,16 +44,30 @@ using namespace lld;
 
 namespace {
-template <class ELFT> class MarkLive {
+
```
delete the blank line after `namespace {`

mysterymath replied:
Done.

MaskRay commented on lld/test/ELF/why-live.test:
```diff
@@ -0,0 +1,160 @@
+
+## Globs match multiple cases. Multiple --why-live flags union.
+
+# RUN: ld.lld a.o a.so --gc-sections --why-live=test_s* | FileCheck %s --check-prefix=MULTIPLE
```
perhaps add another `test_s*` to show the behavior when a symbol is matched by multiple patterns.
We might need `XXX-NOT: live`

mysterymath replied:
Done.

mysterymath committed:
Address review feedback

- Delete blank line after "namespace {"
- Extend glob test to cover multiple pattern matches; add NOT

mysterymath committed:
Add release note

MaskRay commented on lld/test/ELF/why-live.test:
```diff
@@ -0,0 +1,161 @@
+
+## Globs match multiple cases. Multiple --why-live flags union.
+
+# RUN: ld.lld a.o a.so --gc-sections --why-live=test_se* --why-live=test_se* | FileCheck %s --check-prefix=MULTIPLE
```
When `*` is used an an argument, quote it.

mysterymath committed:
Add quotes around wildcards
